### DISCUSSION TITLE
New working design - No Wifi board!!  One ESP32 replaces all of it.

### DISCUSSION THREAD

**Discussion by jazzmonger** [jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip#33] [General]
So someone asked if its possible.  Yup, and i did it.  And the good news is It'll work with the "ultimate" version of my mods, or just the basics version.

It wasn't an easy road.  The ESP32 dev board I started with is frigging defective :(  Only one of the UARTs worked and on 2 of the 40 pins.... sheesh, cheap crap boards from China.  That wasted an entire day and night of fighting pin movement and recompiles... So I replaced it today w/ an ESP32-S2.  Once I mapped out input pins, output pins and UARTs, it all came together and all functions seems to work.

One thing that is very different is that the ADC readings are completely different between the esp8266 and esp32 as they measure voltages differently.  So, I'll have to re-calibrate the exhaust thermocouple sensor.

While I did do a couple of tests w/out using a 5v to 3.3v bidirectional level converter, the safe bet is to use one, so that's what I did.  It's tiny and wont be a problem to wire into place.  In fact, the esp32, level shifter and a new BME280 temp sensor will all fit. If ever I saw a need to create a PC board, this is one of those times!  But for now, I'm pretty good at soldering small wires.  This will be the ultimate test of patience. I might even draw it out first.

![image](https://user-images.githubusercontent.com/52110065/222993064-4857af62-9454-4b61-a389-cab65db6e91d.png)

![8318D694-616F-4E3E-A51A-DB624511261C](https://user-images.githubusercontent.com/52110065/222992698-bf3bd769-b902-45ad-84f4-71cfb7400e81.jpeg)
![881EC127-9395-4607-BE6D-017618EE93B5](https://user-images.githubusercontent.com/52110065/222992718-382c69b3-e607-45b7-921c-d7967ae021b5.jpeg)
![C60BA9FD-7F38-4D68-9C74-A29ADF9B4D9B](https://user-images.githubusercontent.com/52110065/222992725-f3061c77-b96f-4f97-8a80-39b15f2e8031.jpeg)

**Comment by jazzmonger** [comment]
A quick update:
After struggling to get accurate temps from the thermocouple, I decided to try decoding the MCU UART signal that goes TO the display once again.   It turns out that the payload contains 3 temperature sensor readings all at once, room temp, pipe temp and protect temp.  That's it. It was so simple it was embarrassing - I was totally overthinking it. I must have sat and stared at hex for what seems like days.  There are 3 bytes in the middle of each packet payload and each byte contains a temp reading of one of 3 sensors. I have to give credit to @seib on Discord for looking at my output and seeing the pattern of decoding the bytes into the temps shown on the display.  The formula was (hex * convert to C - 34).  Why? Who knows.  There are so many strange things the programmer did on these stoves that I just roll with it at this point.

So,now instead of trying to measure the analog voltage from the exhaust temp thermocouple using the absolutely crappy ADCs on these esps which give ADC converted temps a +/- 30-40 degree swing at any given moment,  I can now get exactly what temp the stove is measuring on all sensors in near real-time. THAT will control the burn perfectly well. The downside is that it takes one more UART (3 total now) to get full functionality in the "ultimate" upgrade. 

UART1- TuyaMCU
UART2-Display message decoding
UART3-main MCU real-time multiple temp decoding.

Normally that wouldn't be an issue as standard ESP32 chips support 3 hardware UARTS, but as I discovered after an entire day of futzing with all the variants (S2,C3,S3,etc), they only support 2 UARTs, and software UARTs are only available on ESP8266 chips.  Guess which ESP32 boards I have? S2 & C3. So, I'm waiting for some ESP32 mini chips to arrive from Amazon.

Once they arrive, I think I'll have all the pieces to finish this new design, completely replacing the Wifi daughter board along with full control over fire pot, etc. Just in time for spring!

**Reply by cpankonien** [reply]
I'm going to have to do a deep dive to see exactly what boards I have. I refer to them as "nodemcus" but I'll have to see exactly what they are, as this sounds like something I gotta try!

**Reply by jazzmonger** [reply]
Amazon says my ESP32 mini boards will be here by 7pm tonight. I'll be working on it today!

And even if you have just a D1 mini, you can use it no problem for a basic Home Assistant integration. In fact, you could get away with no 5v to 3.3v level adjusters as that chip takes 5v inputs according to the Espressif CEO himself. So just wire a D1 mini on a perf board and have at it!

I'm just adding all kinds of bells and whistles that require 3 UARTs from a std ESP32. And as an engineer, I'm designing safety into my ultimate" design bc it controls the Auger. IF you're just sending read-only info to Home assistant, who cares if the D1 fails in a year? just unplug it and replace it!

**Comment by cpankonien** [comment]
I just looked at my Amazon order history, apparently I bought a bunch of these

HiLetgo 2pcs ESP8266 NodeMCU CP2102 ESP-12E Development Board

so they should work, yes?

------- Original Message -------

**Reply by jazzmonger** [reply]
YES, that will work.  It might be tight getting it to fit though as there is very little room beween the socket and back of the case... a D1 mini fits without any problems.  Its running my stove as I type this.

Amazon says my ESP32 mini boards will be here by 7pm tonight. I'll be working on it today!

And even if you have just a D1 mini, you can use it no problem for a basic Home Assistant integration. In fact, you could get away with no 5v to 3.3v level adjusters as that chip takes 5v inputs according to the Espressif CEO himself. So just wire a D1 mini on a perf board and have at it!

I'm just adding all kinds of bells and whistles that require 3 UARTs from a std ESP32. And as an engineer, I'm designing safety into my ultimate" design bc it controls the Auger. IF you're just sending read-only info to Home assistant, who cares if the D1 fails in a year? just unplug it and replace it!

**Reply by jwaz73** [reply]
The D1 Mini works great! That's exactly what I did.  I had some of these D1 Mini clones on hand and thought I would try it without the level shifters.
https://www.amazon.com/gp/product/B07RBNJLK4/ref=ppx_od_dt_b_asin_title_s01?ie=UTF8&psc=1
Just the basic configuration for now but it is working much, much, much better than the original config that came with the stove.  As labeled on the devices I had, D7 goes to RX and D8 goes to TX.  These map to GPIO 13 and 15 respectively.

If you're looking at the back of the display board with the buzzer in the top-left corner, use the connection header that is closest to you.  I'm calling the far-left pin, pin 1.  That is +5v.  Pin 5 is RX and pin 6 is TX.

We are in the middle of moving and I'm at the old house this week.  My stove is at the new house, four hours away...  I want to document this better as soon as this move is done.  Probably just in time to not need the stove for spring.

**Reply by jazzmonger** [reply]
Awesome! Are you using my "basic" firmware file?

Ok, New ESP32 WROOM just arrived and sure enough, they support 3 UARTS
<img width="272" alt="image" src="https://user-images.githubusercontent.com/52110065/225122321-04d310f1-711b-4f01-9bce-b49a9416e11a.png">

I'm wiring it all up now.  I'll post pics of the pins needed on the display and how I repurposed several unused pins.

**Reply by jwaz73** [reply]
Yes, for the most part.  The firmware I used is probably +95% of your efforts (very grateful!).  The changes I made just have to do with HA sensors and the way I monitor WiFi and HA connection status.

I'm looking forward to how you come out with the ESP32 devices.

**Reply by jazzmonger** [reply]
Glad it was useful!  

I got the ESP32-WROOM Mini completely working today on the breadboard. Final wiring continues tomorrow!

**Comment by jazzmonger** [comment]
All installed and working like a champ w/ the D1 Mini ESP32.  I'll upload my config file.
